iree/turbine/kernel/wave/codegen/handlers.py

```diff
@@ -935,49 +935,75 @@ def handle_iterate(emitter: WaveEmitter, node: fx.Node):
 
 
 def handle_iterate_while(emitter: WaveEmitter, node: fx.Node):
-    # TODO: Add support for init args.
     try:
         axis, init_args, subgraph, implicit_capture, start, condition = node.args
     except ValueError as e:
         raise ValidationError("Malformed arguments") from e
 
+    # Flatten init_args and get IR values for each of them.
+    flat_init_args, _ = pytree.tree_flatten((init_args))
+    flat_init_args = [cast_py_value(emitter, arg) for arg in flat_init_args]
+    flat_init_arg_types = [arg.ir_value.type for arg in flat_init_args]
+
     # Initialize while loop
     init_value = cast_py_value(emitter, start).ir_value
     if isinstance(init_value.type, VectorType):
         init_value = vector_d.extractelement(init_value, [0])
-    elif isinstance(init_value.type, IntegerType):
+    if isinstance(init_value.type, IntegerType):
         init_value = arith_d.index_cast(IndexType.get(), init_value)
 
     assert isinstance(
         init_value.type, IndexType
     ), f"Unhandled start type: {init_value.type}"
 
-    whileOp = scf_d.WhileOp([init_value.type], [init_value])
-    whileOp.before.blocks.append(*[init_value.type])
-    whileOp.after.blocks.append(*[init_value.type])
+    init_args = [arg.ir_value for arg in flat_init_args] + [init_value]
+    init_arg_types = flat_init_arg_types + [init_value.type]
+    whileOp = scf_d.WhileOp(init_arg_types, init_args)
+    whileOp.before.blocks.append(*init_arg_types)
+    whileOp.after.blocks.append(*init_arg_types)
 
     # Before block: condition.
-    current_value = whileOp.before.blocks[0].arguments[0]
+    current_values = whileOp.before.blocks[0].arguments
     with InsertionPoint(whileOp.before.blocks[0]):
         # Replace the axis with a temporary variable when generating the condition
         # to avoid conflicts with the actual value of the axis.
         condition = condition.subs({axis: index_symbol("$TMP")})
         subs = add_emitter_subs(emitter)
-        subs[index_symbol("$TMP")] = current_value
+        subs[index_symbol("$TMP")] = current_values[-1]
```

Comment:
any reasons why we index on -1? (as opposed to something like $TMP{i} = current_values[i])

Reply:
is current_values[-1] the condition and the rest are iter_args?

Reply:
Yes, unlike scf.for, scf.while also returns the value of the induction variable. Since we don't expose this, we need to reconcile the difference between what we expose through tkw.iterate and what scf.while does. I am resolving that by keeping the induction variable at the end, so that 0 to num_init_args - 1 is consistent between tkw.iterate and scf.while.

Reply:
makes sense, thanks!

Reply:
can we document/add a comment for this somewhere actually?

Reply:
sure will add in subsequent pr